### ISSUE TITLE
support parsing "set-cookie" header in http/cookie.ts getCookies method

### ISSUE THREAD

**Issue by andykais** [denoland/deno_std#1977]
**Is your feature request related to a problem? Please describe.**
I need to parse the "set-cookie" response header of a request for a web scraper.

**Describe the solution you'd like**
`getCookie` could support parsing "set-cookie" response headers, similar to how it parses "cookie" request headers

**Describe alternatives you've considered**
implementing my own cookie parser logic, since https://deno.land/std/http/cookie.ts doesnt look for the "set-cookie" header

**Comment by getspooky** [comment]
Cookie implementation is based on https://datatracker.ietf.org/doc/html/rfc6265 and doesn't include parsing "set-cookie" response headers.

**Comment by andykais** [comment]
alright. I ended up just reimplementing the `getCookie` logic myself, which is pretty straight forward. If supporting `set-cookie` is out of the scope of std, then thats fine since I have my workaround. Ill let a @getspooky or another contributor close this issue if that is the case.

(side note)
this seems kind of redundant
```ts
    for (const kv of c) {
      const [cookieKey, ...cookieVal] = kv.split("=");
      assert(cookieKey != null);
      const key = cookieKey.trim();
      out[key] = cookieVal.join("=");
    }
```
why not just write
```ts
    for (const kv of c) {
      const [cookieKey, cookieVal] = kv.split("=", 2);
      assert(cookieKey != null);
      const key = cookieKey.trim();
      out[key] = cookieVal;
    }
```

**Comment by getspooky** [comment]
@andykais thanks for feedback, i will check that.

**Comment by jsejcksn** [comment]
> why not just write
> 
> ```ts
>     for (const kv of c) {
>       const [cookieKey, cookieVal] = kv.split("=", 2);
>       assert(cookieKey != null);
>       const key = cookieKey.trim();
>       out[key] = cookieVal;
>     }
> ```

@andykais Setting the [`limit`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#parameters) argument to `2` would exclude all of the parts of the string after a potential second occurrence of `"="`.

Looking at [the syntax in the spec](https://datatracker.ietf.org/doc/html/rfc6265#section-4.1.1), it's not evident to me that `"="` is an invalid token in a cookie value. A safer approach which also avoids creating an intermediate array would be:

```ts
for (const kv of c) {
  const sliceIndex = kv.indexOf("=");
  assert(sliceIndex > 0); // key and value exist, and key length is > 0
  const key = kv.slice(0, sliceIndex).trim();
  out[key] = kv.slice(sliceIndex + 1); // should this be trimmed, too?
}
```

[TS Playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&target=99&useUnknownInCatchVariables=true&exactOptionalPropertyTypes=true#code/PTAEAsBdIBwZwFwgCYFMB2B7AdAGwIbrLByTIACADNgIwBMAHNpcFLMAMaaYDWAlqmyQ4AYgAyNAOySAsACh5aDgQBOqUADMArug6Q+mdKHxw4qFZAAUqAB4wVCUDp5YA7ugA0oALZwA5gD8jqQqfOh+AJSOJmYWcKC29gDciqjK+GqgXOikoJhakI4ASmmYKsgAPCFhfl7V4QB8KXJKqurZuRzBkKHhANoAus3yIKDAAFTjI2Cg46AAChlmWdz8qPGYGsYQqPhoKtOgs6DkMBn43qAA3gASu-twAL47e+bxACrg6l+vKvFhpEIHHUkEwoD8qEgK14AniGhUmG8h2O5DUkC0KiMVwA8gAjABWaUgzzxhL0oFcfEg4Gha1A6Au62M8R4qAAnnBkeNgIdEmUodpdPpDODIQBhVawyw-B6OO6-OBRUAlLjlKo9Gp1DWNa7IlY5KFcGHqAC8Lwe2AhVgARBLjdaIs1Rkc+FtLEa6QBCM3oLS4XARXVyZ1HfW5fKFZWlNX1WqgWMNUBmq6PJ0zUMdQ1J2kCbBwGC4KmWa1JB3NUOaMqgd2GXI8ABueS2HEDV3kFaOmfjheBAEkiLZsw3sGE0DZsRpiyay+2OzFzFY4D3UP2x6BE5RHWMwKy2cYiKB6-hcFp1LY+KQvIRkKBd6BcBg-NTQBf16BKLOK1272bh0u+MCliUHUy6rrYERCKE3iWI6n6hhGfS7gMQ71nmy6WP+fYDjYoAANSgDQW6jHA4D5LgN7Uq+uIglB3ioMgXigpgARwY8epHGiGJGBGaahmxwbppxmLXKmhz8UAA)

**Comment by andykais** [comment]
@jsejcksn oh you are right! I think I have been assuming the limit arg works differently for a while now :sweat_smile: thanks for the tip

**Comment by iuioiua** [comment]
@andykais What's your method of parsing the `set-cookie` header?

**Comment by andykais** [comment]
```ts
function parse_resonse_cookies(response: Response) {
  const cookies: {[key:string]: string} = {}
  for (const [key, value] of response.headers.entries()) {
    if (key === 'set-cookie') {
      const kv_pairs = value
        .split(/;[ ]*/)
        .map(cookie_str => {
          const index = cookie_str.indexOf('=')
          if (index === -1) throw new Error(`invalid cookie '${cookie_str}'`)
          return [
            cookie_str.slice(0, index),
            cookie_str.slice(index + 1)
          ].map(s => s.trim())
        })
      Object.assign(cookies, Object.fromEntries(kv_pairs))
    }
  }
  return cookies
}
```

**Comment by jsejcksn** [comment]
The code in https://github.com/denoland/deno_std/issues/1977#issuecomment-1135331882 won't handle attributes which aren't in the key-value format (e.g. `HttpOnly` and `Secure` — see [`Set-Cookie` on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) for more info): it will throw an exception when encountering those (and they're very common). @iuioiua

I agree that having something in `std` to handle parsing these headers would be nice, however, it's not a trivial task.

I'd like to hear from someone on the Deno core team about whether there's interest in this. If so, I can create a draft PR with an initial implementation to begin the feedback cycle.

**Comment by iuioiua** [comment]
Thank you @andykais, but I did encounter the errors that @jsejcksn mentioned. I decided to go with something far simpler, which suits my use case:
```ts
function getSetCookie(headers: Headers): string {
  return headers.get("set-cookie")!
    .split(", ")
    .map((cookie) => cookie.split("; ")[0])
    .join("; ");
}
```

**Comment by iuioiua** [comment]
> The code in [#1977 (comment)](https://github.com/denoland/deno_std/issues/1977#issuecomment-1135331882) won't handle attributes which aren't in the key-value format (e.g. `HttpOnly` and `Secure` — see [`Set-Cookie` on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) for more info): it will throw an exception when encountering those (and they're very common). @iuioiua
> 
> I agree that having something in `std` to handle parsing these headers would be nice, however, it's not a trivial task.
> 
> I'd like to hear from someone on the Deno core team about whether there's interest in this. If so, I can create a draft PR with an initial implementation to begin the feedback cycle.

@bartlomieju

**Comment by iuioiua** [comment]
@kt3k, this was implemented in #2475.
